File: 2015/python/d20.py
def read(filename):
    for line in open(filename, 'r'):
        return int(line.replace('\n', ''))

def solve(target, per_house, max_presents):
    houses = (target // per_house) + 1
    presents = [0] * houses
    for elf in range(1, houses):
        n = 0
        h = elf
        while (n < max_presents) and (h < houses):
            presents[h] += (per_house * elf)
            n += 1
            h += elf
    for x in range(1, houses):
        if presents[x] >= target:
            return x

input = read('../inputs/input_20.txt')
print(solve(input, 10, input))
print(solve(input, 11, 50))

File: 2015/R/d20.R
d <- as.integer(readLines("../inputs/input_20.txt"))

part1 <- function(d) {
  houses <- rep(0, d/10)
  elf <- 1
  while (elf < d / 10) {
    pick <- seq(elf, d/10, by = elf)
    houses[pick] <- houses[pick] + (10 * elf)
    elf <- elf + 1
  }
  which(houses >= d)[1]
}

part2 <- function(d) {
  houses <- rep(0, d/10)
  elf <- 1
  while (elf < d / 10) {
    pick <- seq(elf, by = elf, length.out = 50)
    houses[pick] <- houses[pick] + (11 * elf)
    elf <- elf + 1
  }
  which(houses >= d)[1]
}
c(part1(d), part2(d))
